unic/utils/src/tables.rs

```diff
@@ -0,0 +1,56 @@
+//! bananas
+
+use std::cmp;
+
+/// bananas
+pub trait CharBsearchTable<V: 'static> {
+    /// bananas
+    fn find(&self, needle: char) -> Option<&V>;
+
+    /// bananas
+    fn binary_search_or<'a>(&'a self, needle: char, default: &'a V) -> &'a V {
+        self.find(needle).unwrap_or(default)
+    }
+
+    /// bananas
+    fn binary_search_or_else<'a>(&'a self, needle: char, f: fn() -> &'a V) -> &'a V {
+        self.find(needle).unwrap_or_else(f)
+    }
+}
+
+impl<V: 'static> CharBsearchTable<V> for &'static [(char, V)] {
+    fn find(&self, needle: char) -> Option<&V> {
+        self.binary_search_by_key(&needle, |&(k, _)| k)
+            .map(|idx| &self[idx].1)
+            .ok()
+    }
+}
+
+impl<V: 'static> CharBsearchTable<V> for &'static [(char, char, V)] {
+    fn find(&self, needle: char) -> Option<&V> {
+        self.binary_search_by(|&(low, high, _)| if needle < low {
+            cmp::Ordering::Less
+        } else if needle > high {
+            cmp::Ordering::Greater
+        } else {
+            cmp::Ordering::Equal
+        }).map(|idx| &self[idx].2)
+            .ok()
+    }
+}
+
+impl CharBsearchTable<()> for &'static [(char, char)] {
+    fn find(&self, needle: char) -> Option<&()> {
+        self.binary_search_by(|&(low, high)| if needle < low {
+            cmp::Ordering::Less
+        } else if needle > high {
+            cmp::Ordering::Greater
+        } else {
+            cmp::Ordering::Equal
+        }).map(|_| {
+                const UNIT: &() = &();
```

Comment:
Why not just return `()` here, so we don't have to have a `const` variable defined and referenced to?

IMHO, returning `Option<_>` from all these endpoints is enough consistency, and there's no need to try to keep reference-ness of the type also be the same. IIRC, there are similar APIs in the std lib, and reference is only returned if there's actual value and it makes sense to do so.

What do you think?